RTT update and Kconfig configuration

nordic-krch opened the pull request:
Pull request contains:
- improvement in RTT console to better handle buffer full situation
- moving RTT configuration from SEGGER_RTT_Conf.h to KConfig
- updating Segger RTT to the latest version

See https://github.com/zephyrproject-rtos/zephyr/pull/8060

nordic-krch committed:
drivers: console: rtt: RTT session awareness

In order to avoid bytes dropping when RTT output buffer is full,
RTT session detection algorithm is introduced to the RTT console.
If host session is detected console retries after wait state
expecting host to empty the buffer. If host session is not detected
console drops the byte without retrying.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
ext: debug: segger: Move RTT configuration to KConfig

Extended KConfig to be able to control configuration present in
SEGGER_RTT_Conf.h file.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
ext: debug: segger: Updating Segger RTT to 6.32d

Segger RTT update includes alignment of SEGGER_RTT_Conf.h to use
KConfig configuration.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
cc: @pizi-nordic

nvlsianpu commented on drivers/console/rtt_console.c:
```diff
@@ -18,14 +18,50 @@
+		 * The host is marked as active if the attempt is successful.
+		 */
+		if (cnt == 0) {
+			if (host_present) {
```
Is it possible to get information of host connection just form the RTT peripheral? (I ques not)

nordic-krch replied:
it's not possible. The only indication is that buffer gets emptied.

MaureenHelm replied:
Maybe reach out to segger to request this as a new feature?

nvlsianpu commented on ext/debug/segger/Kconfig:
```diff
@@ -14,3 +14,45 @@ config SEGGER_SYSTEMVIEW
 	depends on HAS_SEGGER_RTT
 	select RTT_CONSOLE
+
+config SEGGER_RTT_MAX_NUM_UP_BUFFERS
```
It is possible to avoid repetition of 'depends on HAS_SEGGER_RTT' for every entry you added, wrap all entries in `if  HAS_SEGGER_RTT ... endif`  - Do this in your convenience.

MaureenHelm replied:
Agreed. It looks like you made this change, but in the wrong commit. The Kconfig changes should go in the previous commit, not the commit updating RTT to 6.32d.

pizi-nordic commented on drivers/console/Kconfig:
```diff
@@ -143,6 +143,21 @@ config RTT_CONSOLE
 	  Segger J-Link software and displayed on a computer in real-time.
 	  Requires support for Segger J-Link on the companion IC onboard.
 
+config RTT_TX_RETRY_CNT
```
The Host check RTT every dozens of milliseconds (100ms pool rate was frequently observed). The proposed values will not work in such case (algorithm will not detect host). IMHO 10 attempts with 10ms delays between them should catch all the possible cases. @Qbicz: What do you think?

nordic-krch replied:
I've seen 1-2ms period. 10ms of sleep is quite long (i don't know what is the context of this call).

Qbicz replied:
If user is only using RTT up channels (logging from device to PC), the period is usually 1-2 ms as @nordic-krch mentioned. Longer periods may appear if there are frequent RTT writes from PC to device concurrent to RTT reads. But in such case, user will probably easily observe the issue and change this value. For default, I would suggest 2 ms delay x 5 times.

Qbicz replied:
Even during bidirectional usage: logger + console, the polling rate should not exceed few milliseconds. In this scenario, characters typed by user into console (RTT writes PC > device) are not appearing very often and they do not interfere with logger (RTT reads device > PC). The delays in hundreds of milliseconds were only observed when sending larger amounts of data in both RTT directions (tens of kB/s).

nvlsianpu replied:
@Qbicz, @nordic-krch I see that it is not ultra simply to chose appropriate value then. Can you prepare better helper message which explain ho w to do so then? I expect something like - if you observe that expected messages are lost then increase the retries count... This remarks touch also RTT_TX_RETRY_DELAY_MS.

carlescufi commented on drivers/console/rtt_console.c:
```diff
@@ -18,14 +18,63 @@
+static void wait(void)
+{
+	if (k_is_in_isr()) {
+		k_busy_wait(1000*CONFIG_RTT_TX_RETRY_DELAY_MS);
```
spaces: `1000 * CONFIG_RTT_TX_RETRY_DELAY_MS`

carlescufi replied:
Should we really block if we are in an ISR by the way? I wonder if this should be configurable. I guess this will be a moot point once we used the packed stream that comes with the new logging subsystem, but I would still make this configurable so that we don't start getting complaints from users about ISRs running for a long time.

nordic-krch replied:
added option in kconfig

carlescufi commented on drivers/console/rtt_console.c:
```diff
@@ -18,14 +18,63 @@
 	unsigned int key;
 	char c = (char)character;
+	unsigned int cnt;
+	int watchdog_counter = CONFIG_RTT_TX_RETRY_CNT;
```
this is a bit long as a variable name? maybe 'cnt_max'?

nordic-krch replied:
changed.

carlescufi commented on drivers/console/rtt_console.c:
```diff
@@ -18,14 +18,63 @@
+			}
+		} else {
+			/* byte processed - host is present. */
+			host_present = true;
```
You never transition to `false` from `true`, so what happens if the Host disconnects? like for example if the user presses Ctrl+C on `JLinkRTTLogger`

nordic-krch replied:
you will transition to false if cnt equals 0 for number of retries.

MaureenHelm replied:
The logic in this block is really awkward. Is there any way to simplify it? It might help to change the initial condition on line 63 to `if (cnt > 0)` and handle that case first, then move on to the more complicated logic.

carlescufi commented on ext/debug/segger/rtt/SEGGER_RTT_Conf.h:
```diff
@@ -77,22 +78,58 @@ Revision: $Rev: 4351 $
 
-#define SEGGER_RTT_MAX_NUM_UP_BUFFERS             (3)     // Max. number of up-buffers (T->H) available on this target    (Default: 3)
-#define SEGGER_RTT_MAX_NUM_DOWN_BUFFERS           (3)     // Max. number of down-buffers (H->T) available on this target  (Default: 3)
+#define BUFFER_SIZE_UP                            CONFIG_SEGGER_RTT_BUFFER_SIZE_UP        // Size of the buffer for terminal output of target, up to host (Default: 1k)
```
Is there a way we can inject this like we do with mbedTLS or nrfx instead of having to modify the original source from Segger?

nordic-krch replied:
this is configuration file (*_conf.h) so i think it is intended to be modified.

carlescufi commented on ext/debug/segger/Kconfig:
```diff
@@ -8,9 +8,53 @@ config HAS_SEGGER_RTT
+	int "Size of the buffer for terminal input of target, from host"
+	default 16
+
+config SEGGER_RTT_PRINTF_BUFFER_SIZE
```
Do we need this one? I don't think the console uses it?

nordic-krch replied:
it's segger_rtt configuration. Other modules in zephyr may use it as well so it won't harm moving full configuration to kconfig.

carlescufi commented on ext/debug/segger/Kconfig:
```diff
@@ -8,9 +8,53 @@ config HAS_SEGGER_RTT
+config SEGGER_RTT_MODE_NO_BLOCK_TRIM
+	bool "Trim: Do not block, output as much as fits."
+
+config SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL
```
Do we still need to configure this since you've added the `wait()` code in `rtt_console.c`?

nordic-krch replied:
for consistency i think it's worth moving full configuration to kconfig.

carlescufi commented on ext/debug/segger/Kconfig:
```diff
@@ -8,9 +8,53 @@ config HAS_SEGGER_RTT
+
+endchoice
+
+config SEGGER_RTT_USE_ASM
```
Does this work? If it doesn't maybe don't expose it in Kconfig.

nordic-krch replied:
removed

carlescufi commented:
@nordic-krch can you take a look at the comments?

carlescufi commented:
@MaureenHelm @dleach02 could you give this a quick look?

rodrigopex commented:
Hi all, I was trying to use the shell through the RTT console, but RTT_CONSOLE does not implement the input hook (stdin). I was trying to figure out how can I add this, and @carlescufi point me this PR. Is there anything I can do to help to add this feature? My first attempt is to make a polling on the SEGGER_RTT_HasData function to see if there is data to be read from the host if so, write it to the input hook. What do you think? For me, using an ISR would be really better than that, but temporarily, a polling seems to be enough.

nordic-krch commented:
@rodrigopex polling is the only option.

carlescufi commented:
@nordic-krch can you please rebase this PR?
@nvlsianpu can you take another look?

carlescufi reviewed (approved):
LGTM

carlescufi commented:
@nvlsianpu can you please re-review?
@MaureenHelm could you please take a look?

MaureenHelm commented on ext/debug/segger/Kconfig:
```diff
@@ -12,3 +12,45 @@ config SEGGER_SYSTEMVIEW
+	depends on HAS_SEGGER_RTT
+
+choice
+	prompt "Mode for pre-initialized termian channel (buffer 0)"
```
termian?

nvlsianpu commented:
I'm still awaiting for response or/either changes.

nordic-krch commented on drivers/console/rtt_console.c:
```diff
@@ -18,14 +18,65 @@
+		if (cnt == 0) {
+			if (host_present) {
+				wait();
+				max_cnt--;
```
Rephrase the condition. After that change CONFIG_RTT_TX_RETRY_CNT=0 case should be covered.

nordic-krch commented:
@MaureenHelm, I've moved kconfig part to correct commit and refactored loop.
@nvlsianpu, I've addressed your concern regarding help message for tx_retry and delay options.
